### DISCUSSION TITLE
How to pass content of custom control to underlying component control?

### DISCUSSION THREAD

**Discussion by RMPowser** [AvaloniaUI/Avalonia#15906] [Q&A]
So I have a part of my application that needs several instances of a custom control that has a label, a checkbox, and a combobox. I want to be able to write the xaml like this and have it populate the underlying combobox:
```
<cc:CustomTestControl
	Label="BeginSession">
	<ComboBoxItem>Item 1</ComboBoxItem>
	<ComboBoxItem>Item 2</ComboBoxItem>
	<ComboBoxItem>Item 3</ComboBoxItem>
</cc:CustomTestControl>

<cc:CustomTestControl
	Label="EndSession">
	<ComboBoxItem>Other Item 1</ComboBoxItem>
	<ComboBoxItem>Other Item 2</ComboBoxItem>
	<ComboBoxItem>Other Item 3</ComboBoxItem>
	<ComboBoxItem>Other Item 4</ComboBoxItem>
</cc:CustomTestControl>
```

I have tried creating a [TemplateContent] property and then TemplateBinding the ComboBox's ItemsSource property to that, but no luck. I have also tried inheriting from TemplatedControl, ItemsControl, SelectingItemsControl, and ComboBox itself using several implementations and just sort of guessing how this is supposed to be done. I have very limited experience with WPF. Been trying for a several days to figure this out. I have everything else about the custom control working how i want except for this. So frustrating.

**Comment by RMPowser** [comment]
Now I've got it populating the combobox but nothing happens when i click items in the combobox. I can select things with the keyboard by pressing space or enter, but not with the mouse.

Currently im deriving my custom control from SelectingItemsControl. I gave the ComboBox a name of "PART_comboBox" in the axaml. This part is required in the TemplatePart attribute in the code behind. Then i overrode the OnApplyTemplate and used e.NameScope.Find<ComboBox>("PART_comboBox") to get the instance of the combobox. Then i check if its null before setting comboBox.ItemsSource = Items and comboBox.SelectedIndex = 0.

So its populating now, but like i said, clicking any item in the drop down panel doesnt work, and i dont understand why.

**Reply by Pratham95VP** [reply]
It's possible that there's an issue with the focus or mouse capture in your application. You can try setting the` Focusable` property to `true ` on the combobox or its parent elements to see if that helps.

**Reply by RMPowser** [reply]
Hm. No dice. Tried throwing it on every element individually and also all at once. No change in behavior. I noticed something else tho, unrelated to focus.

this is what it looks like when a normal combobox handles PointerPressed and PointerReleased events when not part of a custom control.
![image](https://github.com/AvaloniaUI/Avalonia/assets/53650569/379ba502-2388-471a-a6ec-f8322952d7a2)

and this is what mine looks like. For some reason, it doesnt handle PointerReleased.
![image](https://github.com/AvaloniaUI/Avalonia/assets/53650569/021d4715-8ff3-448d-a2f6-4711476172a2)


To be clear: the first set of Press/Release is for opening the combobox drop down, and the second set is clicking on an element to select it.

**Reply by RMPowser** [reply]
I feel like this should just work the way i have it. Here is the code backing class for the custom control:
```C#
using Avalonia;
using Avalonia.Controls;
using Avalonia.Controls.Metadata;
using Avalonia.Controls.Presenters;
using Avalonia.Controls.Primitives;
using Avalonia.Input;
using Avalonia.LogicalTree;
using Avalonia.Markup.Xaml;
using Avalonia.Metadata;
using System;
using System.Collections;
using System.Collections.ObjectModel;

namespace NYLHostSimulator.CustomControls
{
	[TemplatePart("PART_comboBox", typeof(ComboBox))]
	public partial class CustomTestControl : SelectingItemsControl
	{
		public ComboBox? comboBox;

		public static readonly StyledProperty<string> LabelProperty =
			AvaloniaProperty.Register<CustomTestControl, string>(nameof(Label));

		public string Label
		{
			get { return GetValue(LabelProperty); }
			set { SetValue(LabelProperty, value); }
		}

		protected override void OnApplyTemplate(TemplateAppliedEventArgs e)
		{
			base.OnApplyTemplate(e);

			comboBox ??= e.NameScope.Find<ComboBox>("PART_comboBox");

			if (comboBox != null)
			{
				Items.Insert(0, new ComboBoxItem() { Content = "SRC_OK" });
				comboBox.ItemsSource = Items;
				comboBox.SelectedIndex = 0;
			}
		}
	}
}
```

and here is the xaml:
```xaml
<Styles xmlns="https://github.com/avaloniaui"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:controls="using:NYLHostSimulator.CustomControls">

	<Design.PreviewWith>
		<Border Padding="5">
			<StackPanel>
				
			<controls:CustomTestControl>
				<ComboBoxItem>SRC_OK</ComboBoxItem>
				<ComboBoxItem>lkjshdf;jkhsdf</ComboBoxItem>
				<ComboBoxItem>SRC_OK</ComboBoxItem>
				<ComboBoxItem>SRC_OK</ComboBoxItem>
			</controls:CustomTestControl>
				
			<controls:CustomTestControl>
			</controls:CustomTestControl>
				
			<controls:CustomTestControl>
			</controls:CustomTestControl>
				
			<controls:CustomTestControl>
			</controls:CustomTestControl>
				
			<controls:CustomTestControl>
			</controls:CustomTestControl>
			</StackPanel>
		</Border>
	</Design.PreviewWith>

	<Style Selector="controls|CustomTestControl">

		<Setter Property="Label" Value="Label"/>
		<Setter Property="Template">
				
			<ControlTemplate>
				<Border
					Grid.Row="1"
					Margin="8">

					<DockPanel
						VerticalAlignment="Center">
						<Label
							Padding="0"
							Margin="0 0 0 8"
							DockPanel.Dock="Top"
							VerticalAlignment="Bottom"
							Content="{TemplateBinding Label}">
						</Label>
						
						<DockPanel
							DockPanel.Dock="Bottom">

							<CheckBox
								DockPanel.Dock="Left"
								VerticalAlignment="Top"
								Margin="0 0 16 0">
								Ignore
							</CheckBox>

							<ComboBox
								Name="PART_comboBox"
								DockPanel.Dock="Right"
								VerticalAlignment="Top"
								HorizontalAlignment="Stretch">
							</ComboBox>
						</DockPanel>
					</DockPanel>
				</Border>
			</ControlTemplate>
		</Setter>
	</Style>
</Styles>
```

and here is an example usage scenario:
```xaml
<StackPanel>
        <cc:CustomTestControl
	        Grid.Row="0"
	        Label="BeginSession">
	        <ComboBoxItem>SRC_InvalidInstanceID</ComboBoxItem>
	        <ComboBoxItem>SRC_MustPlayExistingSession</ComboBoxItem>
	        <ComboBoxItem>SRC_NotReadyToPlay</ComboBoxItem>
	        <ComboBoxItem>SRC_SessionInProgress</ComboBoxItem>
	        <ComboBoxItem>SRC_ServiceNotRegistered</ComboBoxItem>
	        <ComboBoxItem>SRC_ServerError</ComboBoxItem>
	        <ComboBoxItem>SRC_PlayCommandNotSent</ComboBoxItem>
        </cc:CustomTestControl>

        <cc:CustomTestControl
	        Grid.Row="1"
	        Label="BeginSessionWithSesstionID">
	        <ComboBoxItem>SRC_InvalidInstanceID</ComboBoxItem>
	        <ComboBoxItem>SRC_InvalidSessionID</ComboBoxItem>
	        <ComboBoxItem>SRC_NotReadyToPlay</ComboBoxItem>
	        <ComboBoxItem>SRC_MustPlayExistingSession</ComboBoxItem>
	        <ComboBoxItem>SRC_OfflineVoucherBarcodeMismatch</ComboBoxItem>
	        <ComboBoxItem>SRC_ServerError</ComboBoxItem>
	        <ComboBoxItem>SRC_SessionIDNotFound</ComboBoxItem>
	        <ComboBoxItem>SRC_SessionInProgress</ComboBoxItem>
	        <ComboBoxItem>SRC_ServiceNotRegistered</ComboBoxItem>
	        <ComboBoxItem>SRC_PlayCommandNotSent</ComboBoxItem>
	        <ComboBoxItem>SRC_SessionEndedVoucherPrintedOffLine</ComboBoxItem>
	        <ComboBoxItem>SRC_SessionNoPreviousSession</ComboBoxItem>
        </cc:CustomTestControl>
</StackPanel
```

**Comment by Pratham95VP** [comment]
```
protected override void OnApplyTemplate(TemplateAppliedEventArgs e)
{
    base.OnApplyTemplate(e);

    comboBox ??= e.NameScope.Find<ComboBox>("PART_comboBox");

    if (comboBox != null)
    {
        Items.Insert(0,  "0_SRC_OK");
        Items.Insert(1, "1_SRC_OK");
        Items.Insert(2, "2_SRC_OK");
        Items.Insert(3, "3_SRC_OK");
        Items.Insert(4, "4_SRC_OK");                
        comboBox.ItemsSource = Items;
        comboBox.SelectedIndex = 0;               
    }
}
```
`Items.Insert` and` Items.Add` are working fine, but something odd is happening with `<ComboBoxItem>`. I haven't had time to figure it out yet because work's been keeping me busy. I'll look into it more after Friday evening.

**Comment by Pratham95VP** [comment]
```
<controls:CustomTestControl>
	<controls:CustomTestControl.Template>
		<ComboBox x:Name="PART_comboBox">
			<ComboBox.Items>SRC_OK</ComboBox.Items>
			<ComboBox.Items>4516</ComboBox.Items>
			<ComboBox.Items>4616</ComboBox.Items>
			<ComboBox.Items>3616</ComboBox.Items>
		</ComboBox>
	</controls:CustomTestControl.Template>
</controls:CustomTestControl>
```

 ```
protected override void OnApplyTemplate(TemplateAppliedEventArgs e)
 {
     base.OnApplyTemplate(e);

     comboBox ??= e.NameScope.Find<ComboBox>("PART_comboBox");

     if (comboBox != null)
     {                
         comboBox.Items.Insert(0, new ComboBoxItem() { Content = "abc5897" });
         comboBox.Items.Insert(1, new ComboBoxItem() { Content = "abc4646" });
          comboBox.SelectedIndex = 0;               
     }
 }
```

Both of these code works for me. 

When you directly add ComboBoxItems to a CustomTestControl in XAML, they show up in the combo box list, but they're not actually added to the ComboBox's Items collection. This happens because the XAML parser treats the ComboBoxItems as part of the CustomTestControl's content, not as items of the ComboBox itself.

The CustomTestControl is a ContentControl, so it can hold various types of content, including ComboBoxItems. However, since the Content property can only hold one item, when you add multiple ComboBoxItems, they get wrapped in a panel (like a StackPanel) to be displayed together.

So, even though the ComboBoxItems are visible in the combo box list, they're not selectable because they're not part of the ComboBox's Items collection

**Reply by RMPowser** [reply]
That makes sense, but then why does the original version work if you replace the ComboBoxItems with system:Strings?
```xaml
<cc:CustomTestControl
	Grid.Row="0"
	Label="BeginSession">
	<system:String>SRC_InvalidInstanceID</system:String>
	<system:String>SRC_MustPlayExistingSession</system:String>
	<system:String>SRC_NotReadyToPlay</system:String>
	<system:String>SRC_SessionInProgress</system:String>
	<system:String>SRC_VLTServiceNotRegistered</system:String>
	<system:String>SRC_ServerError</system:String>
	<system:String>SRC_PlayCommandNotSent</system:String>
</cc:CustomTestControl>
```

**Reply by Pratham95VP** [reply]
If the code works, then there must be something else that I've missed completely. Please provide a minimal sample to investigate this issue like @timunie suggested.

**Comment by timunie** [comment]
If you file a minimum sample one may take a look. In general it should be doable with TemplateBindings

**Reply by RMPowser** [reply]
How should I provide this? Can I just upload a zip with the source right here? Or is there a more proper way?

**Reply by thevortexcloud** [reply]
Github repos are nice since it makes it easier for people to comment on specific lines. But zips also work if you don't have a better option.

**Reply by Pratham95VP** [reply]
Zip file with minimal sample code that demonstrates the issue accompanied by clear comments to help and understand the code.

**Reply by RMPowser** [reply]
I made a quick repo here https://github.com/RMPowser/AvaloniaNestedComboBoxMinSample. should be public

**Reply by timunie** [reply]
I think this doesn't really fly tbh. You could use TemplateBinding on ItemsSource, but that will break Items added directly to your control. If you wanted to extend ComboBox, you should copy over the original style and modify this to your needs.

**Reply by RMPowser** [reply]
So you're saying there's no real way to do this without copying the source code of avalonias ComboBox? If so, this is a problem that inheritance is supposed to solve. So should I inherit from ComboBox instead of SelectingItemsSource? and then somehow inherit from ComboBox's style and extend that?

**Reply by timunie** [reply]
If that is a control that you want to re-use in several projects, yeah that is what I'd do. If it is a thing used loacally only in one App, I'd wire up a UserControl and use normal Bindings, but it would be tied to a ViewModel or Interface of a ViewModel.

**Reply by thevortexcloud** [reply]
>  If so, this is a problem that inheritance is supposed to solve.

Templates can't be inherited. This is a problem WPF has as well. Avalonia does however allow for reusing an existing template by setting the style key, but that is different from inheritance.

https://docs.avaloniaui.net/docs/basics/user-interface/styling/styles#style-key
